File: frontend/inventario-app/services/usuarioService.ts
import { api } from '../api';
import { Usuario } from '../types/usuario';

// 🔹 Probar conexión (listar todos los usuarios)
export const probarConexion = async (): Promise<Usuario[]> => {
  const response = await api.get('/api/usuarios');
  return response.data;
};

// 🔹 Login de usuario
export const loginUsuario = async (cedula: string, contrasena: string): Promise<Usuario> => {
  try {
    const response = await api.post('/api/usuarios/login', { cedula, contrasena });
    return response.data;
  } catch (error: any) {
    console.log(' Error en loginUsuario:', error.response?.data || error.message);

    let mensaje = 'Error al iniciar sesión.';
    if (typeof error.response?.data === 'string') {
      mensaje = error.response.data;
    } else if (error.response?.data?.message) {
      mensaje = error.response.data.message;
    } else if (error.message) {
      mensaje = error.message;
    }

    throw new Error(mensaje);
  }
};

// 🔹 Registrar usuario
export const registrarUsuario = async (usuario: any): Promise<Usuario> => {
  const payload = {
    nombre: usuario.nombre,
    cedula: usuario.cedula,
    correo: usuario.correo,
    contrasena: usuario.contrasena,
    rol: usuario.rol?.toUpperCase() || 'TECNICO',
  };

  console.log(' Enviando usuario a registrar:', payload);
  const response = await api.post('/api/usuarios/registrar', payload, {
    headers: { 'Content-Type': 'application/json' },
  });
  return response.data;
};

// 🔹 Obtener todos los usuarios
export const obtenerUsuarios = async (): Promise<Usuario[]> => {
  const response = await api.get('/api/usuarios');
  return response.data;
};

// 🔹 Cambiar estado (activar / desactivar usuario)
export const cambiarEstadoUsuario = async (idUsuario: number, activo: boolean): Promise<Usuario> => {
  try {
    console.log(` Cambiando estado del usuario ${idUsuario} a: ${activo ? 'Activo' : 'Inactivo'}`);
    const response = await api.put(`/api/usuarios/${idUsuario}/estado`, { activo });
    console.log(' Estado cambiado correctamente:', response.data);
    return response.data;
  } catch (error: any) {
    console.error(' Error al cambiar estado del usuario:', error.response?.data || error.message);
    throw new Error(error.response?.data?.error || 'Error al cambiar estado del usuario.');
  }
};

// 🔹 Actualizar usuario
export const actualizarUsuario = async (
  idUsuario: number,
  nuevosDatos: Partial<Usuario>
): Promise<Usuario> => {
  const response = await api.put(`/api/usuarios/${idUsuario}`, nuevosDatos, {
    headers: { 'Content-Type': 'application/json' },
  });
  return response.data;
};

// 🔹 Obtener usuario por ID
export const obtenerUsuarioPorId = async (id: string | number): Promise<Usuario> => {
  try {
    console.log(` Solicitando usuario con ID: ${id}`);
    const response = await api.get(`/api/usuarios/${id}`);
    console.log(' Usuario obtenido:', response.data);
    return response.data;
  } catch (error: any) {
    console.error(' Error en obtenerUsuarioPorId:', error.response?.data || error.message);
    throw error;
  }
};

File: backend/backend/src/main/java/com/inventario/backend/service/UsuarioService.java
package com.inventario.backend.service;

import com.inventario.backend.utils.ValidadorDatos;
import com.inventario.backend.model.Usuario;
import com.inventario.backend.repository.UsuarioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UsuarioService {

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    // 🔹 Registrar usuario
    public Usuario registrarUsuario(Usuario usuario) {
        if (usuario.getContrasena() == null || usuario.getContrasena().isEmpty()) {
            throw new IllegalArgumentException("La contraseña no puede ser nula o vacía");
        }
        usuario.setContrasena(passwordEncoder.encode(usuario.getContrasena()));
        return usuarioRepository.save(usuario);
    }

    // 🔹 Iniciar sesión
    public Optional<Usuario> login(String cedula, String contrasena) {
        Optional<Usuario> usuarioOpt = usuarioRepository.findByCedula(cedula);

        if (usuarioOpt.isEmpty()) {
            throw new IllegalArgumentException("Cédula no encontrada.");
        }

        Usuario usuario = usuarioOpt.get();

        //  Verificar si está inactivo
     if (!usuario.isActivo()) {
    throw new IllegalArgumentException("El usuario está desactivado. Contacta al administrador.");
}



        //  Validar contraseña
        ValidadorDatos.validarCredenciales(usuario, contrasena);

        return Optional.of(usuario);
    }

    //  Editar perfil
    public Usuario actualizarPerfil(Long idUsuario, Usuario nuevosDatos) {
        Usuario usuario = usuarioRepository.findById(idUsuario)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        //  Si se envía una nueva contraseña, validar todo
        if (nuevosDatos.getContrasena() != null && !nuevosDatos.getContrasena().isEmpty()) {
            ValidadorDatos.validarUsuario(nuevosDatos);
            usuario.setContrasena(passwordEncoder.encode(nuevosDatos.getContrasena()));
        } else {
            // Validar los demás campos, pero ignorar la contraseña
            if (nuevosDatos.getNombre() == null || nuevosDatos.getNombre().trim().isEmpty()) {
                throw new IllegalArgumentException("El nombre no puede estar vacío.");
            }

            if (nuevosDatos.getCedula() == null || !nuevosDatos.getCedula().matches("\\d{6,10}")) {
                throw new IllegalArgumentException("La cédula debe contener entre 6 y 10 dígitos numéricos.");
            }

            if (nuevosDatos.getCorreo() == null ||
                    !nuevosDatos.getCorreo().matches("^[\\w-.]+@[\\w-]+\\.[a-zA-Z]{2,}$")) {
                throw new IllegalArgumentException("El correo electrónico no es válido.");
            }

            if (nuevosDatos.getRol() == null || nuevosDatos.getRol().trim().isEmpty()) {
                throw new IllegalArgumentException("El rol es obligatorio.");
            }

            String rol = nuevosDatos.getRol().toUpperCase();
            if (!rol.equals("ADMIN") && !rol.equals("TECNICO")) {
                throw new IllegalArgumentException("El rol debe ser ADMIN o TECNICO.");
            }
        }

        usuario.setNombre(nuevosDatos.getNombre());
        usuario.setCorreo(nuevosDatos.getCorreo());
        usuario.setRol(nuevosDatos.getRol());
        usuario.setCedula(nuevosDatos.getCedula());

        return usuarioRepository.save(usuario);
    }

    //  Obtener todos los usuarios
    public List<Usuario> obtenerTodos() {
        return usuarioRepository.findAll();
    }

    //  Obtener usuario por id
    public Optional<Usuario> obtenerPorId(Long idUsuario) {
        return usuarioRepository.findById(idUsuario);
    }

    // Cambiar estado del usuario
public Usuario cambiarEstadoUsuario(Long idUsuario, boolean activo) {
    Usuario usuario = usuarioRepository.findById(idUsuario)
            .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

    usuario.setActivo(activo);
    return usuarioRepository.save(usuario);
}

}
